minifier update

lenrik1589 opened the pull request:
- Fixes Avatar loading failing silently if error being thrown is not an `Exception`
- Fixes Long strings in the script crashing the script minifier making avatar not load
- Adds new minifier mode: Abstract Syntaxt Tree minifier (i have no better name for it)

Needs more testing but doesn't crash with the avatars I have.
Contains a temporary mixin into LuaJ AST that fixes handling of non-ascii characters in short strings (needs testing with `[[]]` strings)

lenrik1589 committed:
Loading change
Fix light minifier silently failing with long strings because of the recursive nature of java regex engine and try-catch not catching the error.

Commit:
New minifier mode
- fix `\x00` string escapes
- new minifier mode: Abstract Syntax Tree minifier

Francy-chan commented on src/main/java/org/moon/figura/mixin/StrMixin.java:
since we "own" the lib, those changes should be directly patched in the lib instead of mixin into it

lenrik1589 replied:
Will be resolved by Kingdom-of-The-Moon/luaj#1

Francy-chan commented on src/main/java/org/moon/figura/parsers/LuaScriptBuilderVisitor.java:
this class seems to be worthwhile to be part of the library as well

lenrik1589 replied:
This will have to be here until LuaJ has proper support for concrete syntax trees (parse trees)

Francy-chan commented on src/main/resources/figura.mixins.json:
no mixin into "owned" libraries

lenrik1589 committed:
Add silly stuff
- AST-minified scripts are now marked as garbled line numbers
- added part of what's needed for local variable name mangling

lenrik1589 committed:
Moar silly stuff
- mangle local names

though I'd like variables to have the earlier names the earlier they are used so that it looks nicer instead of having randomly assigned names

may make it sort by number of usages to make it even more optimal, but I'm happy so far

Commit:
Fix vararg exp

Commit:
more fixes
- another round of string fixes
- made `do end` blocks register properly

Commit:
requested changes and more regex
remove LuaJ mixin in favor of modifying the library itself
change pattern modifiers to be localized for multiline string and comment
